File: swift/readline.swift
//******************************************************************************
// MAL - readline
//******************************************************************************

import Foundation

private let HISTORY_FILE = "~/.mal-history"

func with_history_file(do_to_history_file:(UnsafePointer<Int8>) -> ()) {
    HISTORY_FILE.withCString {
        (c_str) -> () in
        let abs_path = tilde_expand(UnsafeMutablePointer<Int8>(c_str))
        if abs_path != nil {
            do_to_history_file(abs_path)
            free(abs_path)
        }
    }
}

func load_history_file() {
    using_history()
    with_history_file {
        let _ = read_history($0)
    }
}

func save_history_file() {
    // Do this? stifle_history(1000)
    with_history_file {
        let _ = write_history($0)
    }
}

func _readline(prompt: String) -> String? {
    let line = prompt.withCString {
        (c_str) -> UnsafeMutablePointer<Int8> in
        return readline(c_str)
    }
    if line != nil {
        if let result = String(UTF8String: line) {
            add_history(line)
            return result
        }
    }
    return nil
}

File: impls/cpp/ReadLine.cpp
#include "ReadLine.h"
#include "String.h"

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

#include <readline/readline.h>
#include <readline/history.h>
#include <readline/tilde.h>

ReadLine::ReadLine(const String& historyFile)
: m_historyPath(copyAndFree(tilde_expand(historyFile.c_str())))
{
    read_history(m_historyPath.c_str());
}

ReadLine::~ReadLine()
{
}

bool ReadLine::get(const String& prompt, String& out)
{
    char *line = readline(prompt.c_str());
    if (line == NULL) {
        return false;
    }
    add_history(line); // Add input to in-memory history
    append_history(1, m_historyPath.c_str());

    out = line;

    return true;
}
